lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,384 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <modem/at_parser.h>
+
+#include "at_token.h"
+#include "at_match.h"
+
+static bool is_crlf(const char *str)
+{
+	return strlen(str) == 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static bool is_subparam(const struct at_token *token)
+{
+	switch (token->type) {
+	case AT_TOKEN_TYPE_INT:
+	case AT_TOKEN_TYPE_QUOTED_STRING:
+	case AT_TOKEN_TYPE_ARRAY:
+	case AT_TOKEN_TYPE_EMPTY:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool is_resp(const char *str)
+{
+	static const char * const resp[] = {
+		"\r\nOK\r\n",
+		"\r\nERROR\r\n",
+		"\r\n+CME ERROR",
+		"\r\n+CMS ERROR"
+	};
+
+	for (size_t i = 0; i < 4; i++) {
+		if (strncmp(str, resp[i], strlen(resp[i])) == 0) return true;
+	}
+
+	return false;
+}
+
+static bool is_index_ahead(struct at_parser *parser, size_t index)
+{
+	return index + 1 > parser->count;
+}
+
+static bool lookahead_crlf_or_more(const char *str)
+{
+	return strlen(str) >= 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static bool lookahead_crlf_and_more(const char *str)
+{
+	return strlen(str) > 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static void at_token_set_empty(struct at_token *token, const char *start)
+{
+	token->start = start;
+	token->len = 0;
+	token->type = AT_TOKEN_TYPE_EMPTY;
+	token->variant = AT_TOKEN_VARIANT_NO_COMMA;
+}
+
+int at_parser_init(struct at_parser *parser, const char *at)
+{
+	if (!parser || !at) {
+		return -EINVAL;
+	}
+
+	size_t len = strlen(at);
+
+	/* Check bounds. Maximum limit of 65535 characters. */
+	if (len == 0 || len > UINT16_MAX) {
+		return -EINVAL;
+	}
+
+	memset(parser, 0, sizeof(struct at_parser));
+
+	parser->cursor = at;
+	parser->is_init = true;
+
+	return 0;
+}
+
+static int at_parser_tok(struct at_parser *parser, struct at_token *token)
+{
+	const char *remainder = NULL;
+	size_t cursor_len = 0;
+	size_t remainder_len = 0;
+	struct at_token tok = {0};
+
+	cursor_len = strlen(parser->cursor);
+
+	/* The lexer cannot match empty strings, so intercept the special case where the empty
+	 * subparameter is the one after the previously parsed token.
+	 * This case is detected in the previous call to this function.
+	 */
+	if (parser->is_next_empty) {
+		at_token_set_empty(&tok, parser->cursor);
+
+		/* Set the remainder to the current cursor because an empty token has length
+		 * zero.
+		 */
+		remainder = parser->cursor;
+
+		/* Reset the flag. */
+		parser->is_next_empty = false;
+
+		goto matched;
+	}
+
+	/* Nothing left to tokenize. */
+	if (cursor_len == 0 || is_crlf(parser->cursor)) {
+		return -EIO;
+	}
+
+	/* Don't tokenize response. */
+	if (is_resp(parser->cursor)) {
+		return -EIO;
+	}
```

Comment:
I responded above, but in essence responses are not used anywhere.

Reply:
but by response you mean `OK` or `ERROR` etc. right? Because when I think about response I think about the whole AT string returned by the modem in response to a command (solicited).